leaf-server/src/main/java/org/dreeam/leaf/async/path/AsyncPath.java

```diff
@@ -78,17 +82,30 @@ public AsyncPath(@NotNull List<Node> emptyNodeList, @NotNull Set<BlockPos> posit
 
     @Override
     public boolean isProcessed() {
-        return this.processState == PathProcessState.COMPLETED;
+        return fastPath || (this.processState == PathProcessState.COMPLETED);
     }
 
     /**
      * returns the future representing the processing state of this path
      */
-    public synchronized void postProcessing(@NotNull Runnable runnable) {
+    public void postProcessing(@NotNull Runnable runnable) {
         if (isProcessed()) {
             runnable.run();
-        } else {
-            this.postProcessing.add(runnable);
+            return;
+        }
+
+        boolean finished = this.processingFinished;
+        if (finished) {
+            runnable.run();
+            return;
+        }
+
+        synchronized (this) {
+            if (isProcessed()) {
+                runnable.run();
+            } else {
+                this.postProcessing.add(runnable);
+            }
         }
```

Comment:
I think you misunderstood my previous comment, I meant this:

```suggestion
        boolean shouldRun = false;
        synchronized (this) {
            if (isProcessed()) {
                shouldRun = true;
            } else {
                this.postProcessing.add(runnable);
            }
        }
        if (shouldRun) {
            runnable.run();
        }
```

Reply:
This pulls the runnable out of the critical section